homeassistant/components/sensor/swiss_public_transport.py

```diff
@@ -44,39 +48,39 @@
 })
 
 
-def setup_platform(hass, config, add_devices, discovery_info=None):
+@asyncio.coroutine
+def async_setup_platform(hass, config, async_add_devices, discovery_info=None):
     """Set up the Swiss public transport sensor."""
     name = config.get(CONF_NAME)
-    # journal contains [0] Station ID start, [1] Station ID destination
-    # [2] Station name start, and [3] Station name destination
-    journey = [config.get(CONF_START), config.get(CONF_DESTINATION)]
-    try:
-        for location in [config.get(CONF_START), config.get(CONF_DESTINATION)]:
-            # transport.opendata.ch doesn't play nice with requests.Session
-            result = requests.get(
-                '{}locations?query={}'.format(_RESOURCE, location), timeout=10)
-            journey.append(result.json()['stations'][0]['name'])
-    except KeyError:
-        _LOGGER.exception(
-            "Unable to determine stations. "
-            "Check your settings and/or the availability of opendata.ch")
+    start = config.get(CONF_START)
+    destination = config.get(CONF_DESTINATION)
+
+    connection = SwissPublicTransportSensor(hass, start, destination, name)
+    yield from connection.async_update()
+
+    if connection.state is None:
+        _LOGGER.error(
+            "Check at http://transport.opendata.ch/examples/stationboard.html "
+            "if your station names are valid")
         return False
 
-    data = PublicTransportData(journey)
-    add_devices([SwissPublicTransportSensor(data, journey, name)], True)
+    async_add_devices([connection])
 
 
 class SwissPublicTransportSensor(Entity):
     """Implementation of an Swiss public transport sensor."""
 
-    def __init__(self, data, journey, name):
+    def __init__(self, hass, start, destination, name):
         """Initialize the sensor."""
-        self.data = data
+        from opendata_transport import OpendataTransport
+
+        self.hass = hass
         self._name = name
-        self._state = None
-        self._times = None
-        self._from = journey[2]
-        self._to = journey[3]
+        self._from = start
+        self._to = destination
+        self._websession = async_get_clientsession(self.hass)
+        self._opendata = OpendataTransport(
```

Comment:
Why not create the `OpendataTransport` instance in `async_setup_platform` and pass it as `data` to the init here? You wouldn't need to pass in `hass` then. `self._websession` is not used on this `Entity` instance so that also seems misplaced.